### PR TITLE
`HirId` to `LocalDefId` cleanup

### PR DESCRIPTION
**Pull request by lcnr**
revival of the still relevant parts of #109125

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 1]
```diff
@@ -1116,7 +1115,7 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
     // FIXME(vincenzopalazzo): move the HirId to a LocalDefId
```
Comment (deltragon) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:1115]: This fixme is outdated now.

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 2]
```diff
@@ -212,7 +212,7 @@ pub trait TypeErrCtxtExt<'tcx> {
 
     fn extract_callable_info(
         &self,
-        hir_id: HirId,
+        hir_id: LocalDefId,
```
Comment (deltragon) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:215]:
```suggestion
        body_id: LocalDefId,
```

**Review by deltragon — APPROVED** [review]

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 3]
```diff
@@ -1168,7 +1166,7 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                     })
                 }
                 ty::Param(param) => {
-                    let generics = self.tcx.generics_of(hir_id.owner.to_def_id());
+                    let generics = self.tcx.generics_of(body_id);
```
Comment (cjgillot) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:1169]: This is not exactly equivalent for closures, but the difference doesn't matter.
r=me once CI passes

**Pull request merged**